core/src/transaction_weighting_stage.rs

```diff
@@ -0,0 +1,77 @@
+use {
+    crossbeam_channel::{Receiver, RecvTimeoutError, Sender},
+    rayon::prelude::*,
+    solana_gossip::cluster_info::ClusterInfo,
+    solana_perf::packet::PacketBatch,
+    solana_runtime::bank_forks::BankForks,
+    solana_streamer::streamer::{self, StreamerError},
+    std::{
+        collections::HashMap,
+        net::IpAddr,
+        sync::{Arc, RwLock},
+        thread::{self, Builder, JoinHandle},
+        time::Instant,
+    },
+};
+
+pub struct TransactionWeightStage {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl TransactionWeightStage {
+    pub fn new(
+        packet_receiver: Receiver<PacketBatch>,
+        sender: Sender<Vec<PacketBatch>>,
+        bank_forks: Arc<RwLock<BankForks>>,
+        cluster_info: Arc<ClusterInfo>,
+    ) -> Self {
+        let thread_hdl = Builder::new()
+            .name("sol-tx-weight".to_string())
+            .spawn(move || {
+                let mut last_stakes = Instant::now();
+                let mut ip_to_stake: HashMap<IpAddr, u64> = HashMap::new();
+                loop {
+                    if last_stakes.elapsed().as_millis() > 1000 {
+                        let root_bank = bank_forks.read().unwrap().root_bank();
+                        let staked_nodes = root_bank.staked_nodes();
+                        ip_to_stake = cluster_info
+                            .tvu_peers()
+                            .into_iter()
+                            .filter_map(|node| {
+                                let stake = staked_nodes.get(&node.id)?;
+                                Some((node.tvu.ip(), *stake))
+                            })
+                            .collect();
+                        last_stakes = Instant::now();
+                    }
+                    match streamer::recv_packet_batches(&packet_receiver) {
+                        Ok((mut batches, _num_packets, _recv_duration)) => {
+                            Self::apply_weights(&mut batches, &ip_to_stake);
+                            if let Err(e) = sender.send(batches) {
+                                info!("Sender error: {:?}", e);
+                            }
+                        }
+                        Err(e) => match e {
+                            StreamerError::RecvTimeout(RecvTimeoutError::Disconnected) => break,
+                            StreamerError::RecvTimeout(RecvTimeoutError::Timeout) => (),
+                            _ => error!("error: {:?}", e),
+                        },
+                    }
+                }
+            })
+            .unwrap();
+        Self { thread_hdl }
+    }
+
+    fn apply_weights(batches: &mut [PacketBatch], ip_to_stake: &HashMap<IpAddr, u64>) {
+        batches.into_par_iter().for_each(|batch| {
+            batch.packets.par_iter_mut().for_each(|packet| {
```

Comment:
1. common practice throughout the code base is to have a dedicated threadpool for these rayon calls: https://github.com/solana-labs/solana/blob/8b66625c957f0e44598e1ddf647a9b5733946fca/ledger/src/blockstore.rs#L74-L78
2. Can we use rayon's `flat_map` to break this iterator of iterators into a single parallel iterator? Threads will get more even distribution of work that way if some of these batches are bigger than others